### ISSUE TITLE
Randomly gives incorrect solution when using the OSQP solver

### ISSUE THREAD

**Issue by philip4096** [cvxgrp/cvxpy#620]
I have installed 1.0.10 on windows using pip with python 3.7. ```nosetests cvxpy``` gives no errors.
The second assert in the below example fails randomly after 0-5 times, giving output similar to this.
```
AssertionError: 
Not equal to tolerance rtol=0.01, atol=0
OSQP failed 1
(mismatch 100.0%)
 x: array([ 0.070337, -0.010408])
 y: array([ 0.088796, -0.094413])
```
No problems with ECOS.

```
import numpy as np
import cvxpy

x = cvxpy.Variable(2)
x0 = cvxpy.Parameter(2)
x1 = cvxpy.Parameter(2)
x2 = cvxpy.Parameter(2)

C = np.diag([100.0, 10.0])

D = np.array([[12.07907757, 1.30104775],
              [ 1.30104775,  3.5497937 ]])

E = np.array([[2., 1.],
              [1., 2.]])

obj = - cvxpy.quad_form(x - x0, C) + \
      - cvxpy.quad_form(x - x1, D) + \
      - cvxpy.quad_form(x - x2, E)

prob = cvxpy.Problem(cvxpy.Maximize(obj))

x0val = np.array([.08, 0.00])
x0.value = x0val

x1val = np.array([0.0, 0.0])
x1.value = x1val

x2.value = x1val

xcorrect = np.linalg.solve(C + D + E, (D + E) @ x1val + C @ x0val)

for i in range(100):
    prob.solve(solver ='ECOS', abstol = 1e-12, reltol = 1e-12)
    np.testing.assert_allclose(xcorrect, x.value, rtol = 1e-5, err_msg="ECOS failed")

for i in range(100):
    prob.solve(solver ='OSQP', eps_abs = 1e-12, eps_rel = 1e-12)
    np.testing.assert_allclose(xcorrect, x.value, rtol=1e-2, err_msg=f"OSQP failed {i}")

```

**Comment by gbanjac** [comment]
PR https://github.com/cvxgrp/cvxpy/pull/621 should fix this issue.

I realized that some data change across multiple solves in the example above. This is not related to the OSQP interface though, but the way CVXPY builds the problem data.

@SteveDiamond This should not be the expected behavior, right?

**Comment by SteveDiamond** [comment]
Certainly the solver shouldn't succeed and then fail. Right now cvxpy rebuilds the problem after each solve, and checks if the data changed from the previous solve, updating if necessary.

**Comment by bstellato** [comment]
Apart from this problem that should be fixed in #621, I think that cvxpy variables and constraints do not preserve the order between different solves. Am I right? 

If the problem is rebuilt it might have constraints/variables in different orders which correspond to QP matrices with permuted rows/columns. This makes it harder to detect if OSQP (or any other solver) problem updates can be applied correctly.

One idea could be to force the variables and constraints IDs to be in increasing/decreasing order so that the problem rows/columns are not permuted. What do you think?

**Comment by gbanjac** [comment]
@bstellato Exactly. Removing randomness from the data building process would make warm starting much more efficient. At the moment, OSQP needs to refactor the KKT matrix whenever the ordering of variables or constraints is not preserved.
